### PR TITLE
feat(storage): test for xor filter

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

test for xor filter

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
#6393

### CODE REVIEW

**Comment by soundOfDestiny** [comment]
bench_bloom_filter_read time:   [5.2925 µs 5.4760 µs 5.6653 µs]
                        change: [-12.211% -8.4005% -4.5495%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

===bench_bloom_filter_read fpr===: 0.06761233229399816%
bench_xor_filter_read   time:   [2.3530 µs 2.3792 µs 2.4067 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

===bench_xor_filter_read fpr===: 0.23188193748638772%

c.c. @hzxa21 @Little-Wallace

**Review by hzxa21 — COMMENTED** [review]
We should compare FPR and memory consumption together since less bits per key normally results in larger FPR for a given probabilistic data structure. We can compare Xor、Fuse (an improvement on Xor: https://arxiv.org/abs/2201.01174) and Ribbon (an variant of Xor).

At the first glance, Binary Fuse filters seems to be promising with lower memory consumption, smaller FPR and acceptable query speed. I suggest we do more investigation on that.

File: src/storage/src/hummock/sstable/mod.rs [hunk 1]
```diff
@@ -160,6 +160,12 @@ impl Sstable {
         table_id.bitxor(dist_key_hash)
     }
 
+    #[inline(always)]
+    pub fn hash_for_bloom_filter_u64(dist_key: &[u8], table_id: u32) -> u64 {
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/mod.rs:164]: We need some comments to expain the context of table_id XOR

**Comment by soundOfDestiny** [comment]
===bench_bloom_filter_read bpe===: 20.0000008
===bench_bloom_filter_read fpr===: 0.008329391656716001%
===bench_xor8_filter_read fpr===: 0.39061%
===bench_fuse8_filter_read bpe===: 9.0177536
===bench_fuse8_filter_read fpr===: 0.39044%
===bench_fuse16_filter_read bpe===: 18.0355072
===bench_fuse16_filter_read fpr===: 0.0016%

I think we can migrate to fuse16 @hzxa21  @Little-Wallace

**Comment by soundOfDestiny** [comment]
bench_bloom_filter_read time:   [5.6514 µs 5.7701 µs 5.8998 µs]
                        change: [-22.704% -18.857% -15.161%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

===bench_bloom_filter_read bpe===: 14.00048828125
bench_xor_filter_read   time:   [3.0219 µs 3.1542 µs 3.3119 µs]
                        change: [+13.443% +18.110% +23.331%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 18 outliers among 100 measurements (18.00%)
  3 (3.00%) low mild
  13 (13.00%) high mild
  2 (2.00%) high severe

bench_fuse8_filter_read time:   [2.6341 µs 2.7141 µs 2.7993 µs]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe

===bench_fuse8_filter_read bpe===: 10
bench_fuse16_filter_read
                        time:   [2.9015 µs 3.1115 µs 3.4117 µs]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

===bench_fuse16_filter_read bpe===: 20

**Comment by soundOfDestiny** [comment]
xor(not fuse) result:
bench_bloom_filter_read time:   [5.9050 µs 6.0102 µs 6.1202 µs]
                        change: [+6.1475% +8.9961% +11.905%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

===bench_bloom_filter_read bpe===: 14.00048828125
bench_xor8_filter_read  time:   [2.9168 µs 2.9554 µs 2.9905 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild

===bench_xor8_filter_read bpe===: 9.85546875
bench_xor16_filter_read time:   [2.7624 µs 2.8096 µs 2.8602 µs]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

===bench_xor16_filter_read bpe===: 19.7109375

**Comment by soundOfDestiny** [comment]
xor(not fuse) result:
===bench_bloom_filter_read bpe===: 20.0000008
===bench_bloom_filter_read fpr===: 0.008329391656716001%
===bench_xor8_filter_read bpe===: 9.840024
===bench_xor8_filter_read fpr===: 0.39061%
===bench_xor16_filter_read bpe===: 19.680048
===bench_xor16_filter_read fpr===: 0.00133%

**Comment by soundOfDestiny** [comment]
#7834 is now merged

**Pull request closed without merging**